File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/interactor/BookListAuthorSorter.java
package com.example.sergejmitrofanov.librarycatalogue.interactor;

import androidx.annotation.NonNull;
import com.example.sergejmitrofanov.librarycatalogue.entity.Book;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;

public class BookListAuthorSorter implements BookListSorter {

  @Inject
  public BookListAuthorSorter() {
  }

  @SuppressWarnings("ComparatorCombinators")
  @NonNull
  @Override
  public List<Book> sortBooks(@NonNull List<Book> books) {
    List<Book> sortedBooks = new ArrayList<>(books);
    Collections.sort(sortedBooks, (o1, o2) -> o1.author.compareTo(o2.author));
    return sortedBooks;
  }
}

File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/interactor/BookListAuthorSorter.kt
package com.example.sergejmitrofanov.librarycatalogue.interactor

import com.example.sergejmitrofanov.librarycatalogue.entity.Book
import javax.inject.Inject

class BookListAuthorSorter @Inject constructor() : BookListSorter {
    override fun sortBooks(books: List<Book>) = books.sortedWith { b1: Book, b2: Book ->
        val result = b1.author.compareTo(b2.author)
        if (result == 0) {
            b1.bookTitle.compareTo(b2.bookTitle)
        } else {
            result
        }
    }
}
